In [37]:
import pandas as pd

In [38]:
df = pd.read_csv("https://raw.githubusercontent.com/jakevdp/data-CDCbirths/master/births.csv")
print('Dataframe shape:', df.shape)


Dataframe shape: (15547, 5)


In [39]:
print(df.columns.tolist())

['year', 'month', 'day', 'gender', 'births']


In [40]:
df.gender.value_counts()

F    7776
M    7771
Name: gender, dtype: int64

In [41]:
df.head() #first five observations

year  month  day gender  births
0  1969      1  1.0      F    4046
1  1969      1  1.0      M    4440
2  1969      1  2.0      F    4454
3  1969      1  2.0      M    4548
4  1969      1  3.0      F    4548

In [42]:
df.tail() #last 5 observations

year  month  day gender  births
15542  2008     10  NaN      M  183219
15543  2008     11  NaN      F  158939
15544  2008     11  NaN      M  165468
15545  2008     12  NaN      F  173215
15546  2008     12  NaN      M  181235

We see that that the data is relatively simple–it contains the number of births grouped by date and gender. The tail values show some missing values. Let's investigate further.

In [43]:
df.describe()

year         month           day         births
count  15547.000000  15547.000000  15067.000000   15547.000000
mean    1979.037435      6.515919     17.769894    9762.293561
std        6.728340      3.449632     15.284034   28552.465810
min     1969.000000      1.000000      1.000000       1.000000
25%     1974.000000      4.000000      8.000000    4358.000000
50%     1979.000000      7.000000     16.000000    4814.000000
75%     1984.000000     10.000000     24.000000    5289.500000
max     2008.000000     12.000000     99.000000  199622.000000

In [44]:
df.isnull().sum()

year        0
month       0
day       480
gender      0
births      0
dtype: int64

There are 480 missing values for the day of birth. Also note an error in the max of day (currently 99) which should not be more than 31 depending on the month of the year. The std is also relatively high due to anomalously high values.

In [45]:
df[df["day"]>31] #Months with more than 31 days

year  month   day gender  births
62     1969      1  99.0      F      26
63     1969      1  99.0      M      38
126    1969      2  99.0      F      42
127    1969      2  99.0      M      48
190    1969      3  99.0      F      64
...     ...    ...   ...    ...     ...
14572  1988      4  99.0      F       1
14635  1988      5  99.0      F       1
14696  1988      6  99.0      F       1
14697  1988      6  99.0      M       1
14944  1988     10  99.0      M       1

[350 rows x 5 columns]

In [62]:
list_of_day = df['day'].tolist() #prefer changing to list
day_np=np.array(list_of_day) #modify to np array
day_np[day_np > 31] = np.nan #find months that indicate day above 31 and replace with nan
#check if months 1,2,3,5,7,8,10,12 have 31 days or less
#check if months 9,4,6 and 11 have days equal or less than 30
day_mean=np.nanmean(day_np)
day_mean
#df['day'].replace(df[df["day"]>31], np.NaN)

15.838078412719984

In [47]:
mean_day = df['day'].mean()
mean_day

17.769894471361255

In [22]:
df.isnull().sum()

year        0
month       0
day       480
gender      0
births      0
dtype: int64

In [27]:
df.describe()

year         month         births
count  15547.000000  15547.000000   15547.000000
mean    1979.037435      6.515919    9762.293561
std        6.728340      3.449632   28552.465810
min     1969.000000      1.000000       1.000000
25%     1974.000000      4.000000    4358.000000
50%     1979.000000      7.000000    4814.000000
75%     1984.000000     10.000000    5289.500000
max     2008.000000     12.000000  199622.000000